### PR TITLE
Validation

### PR DESCRIPTION
**Pull request by Rahaf-96**
- Sign up Validation onback end side using joi.
- Check if email exists in database.
- Check if email domain is @gazaskygeeks.com

### CODE REVIEW

File: server/controllers/postUser.js [hunk 1]
```diff
@@ -1,16 +1,43 @@
 const postUserQuery = require("../database/queries/addUsers");
+
+const joi = require("@hapi/joi");
+const validationObject = {
+  username: joi // username should be string and at least it contains 2 characters
+    .string()
+    .min(2)
+    .max(40),
```
Comment (ghassanmas) [inline server/controllers/postUser.js:8]: So here you are saying username can't be more than 40 characters, which is okay but its not connsisient with the DB constraint where you set the username as 50 characters: https://github.com/GSG-K3/code-academy-wiki/blob/master/server/database/config/dbBuild.sql#L67-L74

File: server/controllers/postUser.js [hunk 2]
```diff
@@ -1,16 +1,43 @@
...
-  postUserQuery(userInfo);
-  res.status(200).json({ message: "user  created succefully" }); 
-}
- catch(err){
-      res.status(400).json({message:"There is no valid information"})
+  const { error, value } = schema.validate(req.body);
+  const { username, email, password } = value;
+  if (error) {
+    res.status(400).json({ error: error.details[0].message });
+  } else {
```
Comment (ghassanmas) [inline server/controllers/postUser.js:30]: I think here you should check for if `schema.validate` returned an error before doing destructuring for `value`. There is no point of doing destructing if there is an error.

File: server/controllers/postUser.js [hunk 3]
```diff
@@ -1,43 +1,45 @@
...
+        } else {
+          res
+            .status(400)
+            .json({ error: 'email domain must be gazaskygeeks.com' });
+        }
+      } else {
+        res.status(400).json({ message: 'email already exists!' });
+      }
+    })
+    .catch((err) => err);
```
Comment (ghassanmas) [inline server/controllers/postUser.js:44]: So your code is workable but definitely not maintainable nor is it readable.  I know you are short in time and so... still try your best to abstract it hence you are in **reactor week**.
I will give a hint or a little guide to do so... for example try to be consistent between using `async await` or `promise then`
Suppose your choice is `async await`, here is how a better code structure would be: 


```
try{
//check if email already exist if not throw an error with specific type/number/code
//validate data from client if not throw an error corresponding to the type of error
// insert the user if not successful throw an error with specific type/number/code
//res.send("succes...etc..)
}
catch(e){
 let error;
switch(e.code){
 case 1:
error="Email already exist");
break;
.....etc
}
res.state(400).json({message:error})
}
````

Comment (ghassanmas) [reply server/controllers/postUser.js:44]: you can helper function/s to help you get you to the level of abstraction mentioned above

Comment (Rahaf-96) [reply server/controllers/postUser.js]: @ghassanmas I refactored the code using middlewares and helpers. Is it better now?

File: server/controllers/postUser.js [hunk 4]
```diff
@@ -1,43 +1,21 @@
...
+  // store user information to datbase
+  try {
+    postUserQuery(req.body, (postError) => {
+      if (postError) {
+        return res.status(400).json({ message: postError.message });
+      }
+      return res.status(200).json({ message: 'user  created successfully' });
+    });
+  } catch (err) {
+    res.status(400).json({ message: 'There is no valid information' });
```
Comment (ghassanmas) [inline server/controllers/postUser.js:19]: you can use `async await` instead of callback for for the `postUserQuery` hence you already using try catch....

Comment (Rahaf-96) [reply server/controllers/postUser.js]: @ghassanmas I've changed it, could you check please if I did it in the right way?

**Pull request merged**